Disable MQTT devservice for tests

Hello, 

we want to use your HiveMQ client in a current project, but we have the problem with developers who work on the Mac that HiveMQ does not work under the M4.
For the development we run Quarkus via Docker Compose and here you can also set the platform for the HiveMQ service to `linux/amd64`, then the HiveMQ container starts. 

We did not manage to do this for the tests. Therefore, one solution would be to deactivate the devservice via the configuration variable `quarkus.mqtt.devservices.enabled=false`. Unfortunately this is ignored, when I - admittedly - took a cursory look at your implementation I saw that you only check whether the devservices are deactivated in general and not whether the service is also deactivated via `quarkus.mqtt.devservices.enabled=false`.

Even setting SmallRey InMemory did not help: 
`%test.mp.messaging.outgoing.[channel].connector=smallrye-in-memory`

Is there any hope for improvement here?


https://github.com/quarkiverse/quarkus-hivemq-client/blob/db019ea04ab4fc3357cf0de9dde6005f0c77a4e9/deployment/src/main/java/io/quarkiverse/hivemqclient/deployment/MqttDevServicesProcessor.java#L137


Thx and greeting 
Rico

## dependencies: 
    implementation("$quarkusGroupId:quarkus-messaging-mqtt")
    implementation("io.quarkiverse.hivemqclient:quarkus-hivemq-client:2.4.0")

## config:
```

%test.quarkus.mqtt.devservices.enabled=false
%test.mp.messaging.outgoing.out-box.connector=smallrye-in-memory
```

## Error:
```
java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkiverse.hivemqclient.deployment.MqttDevServicesProcessor#startMqttDevService threw an exception: java.lang.RuntimeException: org.testcontainers.containers.ContainerLaunchException: Container startup failed for image hivemq/hivemq4:latest
	at io.quarkiverse.hivemqclient.deployment.MqttDevServicesProcessor.startMqttDevService(MqttDevServicesProcessor.java:95)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:735)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:255)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
	at java.base/java.lang.Thread.run(Thread.java:1575)
	at org.jboss.threads.JBossThread.run(JBossThread.java:499)
Caused by: org.testcontainers.containers.ContainerLaunchException: Container startup failed for image hivemq/hivemq4:latest
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:351)
	at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:322)
	at io.quarkiverse.hivemqclient.deployment.MqttDevServicesProcessor.lambda$startMqttBroker$1(MqttDevServicesProcessor.java:164)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at io.quarkiverse.hivemqclient.deployment.MqttDevServicesProcessor.startMqttBroker(MqttDevServicesProcessor.java:182)
	at io.quarkiverse.hivemqclient.deployment.MqttDevServicesProcessor.startMqttDevService(MqttDevServicesProcessor.java:79)
	... 10 more
Caused by: org.rnorth.ducttape.RetryCountExceededException: Retry limit hit with exception
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:88)
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:336)
	... 15 more
Caused by: org.testcontainers.containers.ContainerLaunchException: Could not create/start container
	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:556)
	at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:346)
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81)
	... 16 more
```

--- a/deployment/src/main/java/io/quarkiverse/hivemqclient/deployment/MqttDevServicesProcessor.java
+++ b/deployment/src/main/java/io/quarkiverse/hivemqclient/deployment/MqttDevServicesProcessor.java
@@ -218,12 +218,18 @@ public class MqttDevServicesProcessor {
 
     private MqttDevServiceCfg getConfiguration() {
         Config config = ConfigProvider.getConfig();
-        boolean devServicesEnabled = config.getOptionalValue("hivemq.devservices.enabled", Boolean.class).orElse(true);
-        String imageName = config.getOptionalValue("hivemq.devservices.image-name", String.class).orElse("hivemq/hivemq4");
-        Integer fixedExposedPort = config.getOptionalValue("hivemq.devservices.port", Integer.class).orElse(0);
-        boolean shared = config.getOptionalValue("hivemq.devservices.shared", Boolean.class).orElse(false);
-        String serviceName = config.getOptionalValue("hivemq.devservices.service-name", String.class).orElse("mqtt");
+        boolean globalDevServiceEnabled = config.getOptionalValue("quarkus.devservices.enabled", Boolean.class).orElse(true);
+        boolean mqttDevServiceEnabled = config.getOptionalValue("quarkus.hivemq.devservices.enabled", Boolean.class)
+                .orElse(true);
+
+        boolean devServicesEnabled = globalDevServiceEnabled && mqttDevServiceEnabled;
+        String imageName = config.getOptionalValue("quarkus.hivemq.devservices.image-name", String.class)
+                .orElse("hivemq/hivemq4");
+        Integer fixedExposedPort = config.getOptionalValue("quarkus.hivemq.devservices.port", Integer.class).orElse(0);
+        boolean shared = config.getOptionalValue("quarkus.hivemq.devservices.shared", Boolean.class).orElse(false);
+        String serviceName = config.getOptionalValue("quarkus.hivemq.devservices.service-name", String.class).orElse("mqtt");
         Map<String, String> containerEnv = new HashMap<>();
+
         return new MqttDevServiceCfg(new MqttDevServicesBuildTimeConfig(devServicesEnabled, imageName, fixedExposedPort, shared,
                 serviceName, containerEnv));
     }

--- a/docs/modules/ROOT/pages/index.adoc
+++ b/docs/modules/ROOT/pages/index.adoc
@@ -174,4 +174,27 @@ for example, you can run `mvn -V -B -am clean verify -Dnative -Dquarkus.native.b
 
 == Dev Services
 
-If no `outgoing/incoming host/port` is defined then Dev Services will launch a default instance and will connect your channels to the HiveMQ broker.
+The **Dev Services** feature in the HiveMQ Quarkus extension simplifies the integration of a local MQTT broker in your application for testing and development. If no custom host/port configuration is defined, Dev Services will automatically launch an instance of the MQTT broker and connect your messaging channels to it.
+
+The configuration for Dev Services can be controlled via the `application.properties` file. The following properties are available for customization:
+
+[source,properties]
+----
+# Enable/Disable DevServices globally
+quarkus.devservices.enabled=true
+
+# Enable/Disable MQTT DevServices specifically
+quarkus.hivemq.devservices.enabled=true
+
+# Set the image name for the MQTT broker (defaults to 'hivemq/hivemq4')
+quarkus.hivemq.devservices.image-name=hivemq/hivemq4
+
+# Set a fixed port for the MQTT broker (set to 0 to let Docker pick an available port)
+quarkus.hivemq.devservices.port=0
+
+# Configure whether the DevService should be shared across multiple tests
+quarkus.hivemq.devservices.shared=false
+
+# Set a custom service name for the DevService container
+quarkus.hivemq.devservices.service-name=mqtt
+----

--- a/integration-tests/hivemq-client-smallrye/src/main/resources/application.properties
+++ b/integration-tests/hivemq-client-smallrye/src/main/resources/application.properties
@@ -16,3 +16,6 @@ mp.messaging.incoming.prices.auto-generated-client-id=true
 mp.messaging.incoming.custom-topic-sink.connector=smallrye-mqtt-hivemq
 mp.messaging.incoming.custom-topic-sink.topic=custom-topic
 mp.messaging.incoming.custom-topic-sink.auto-generated-client-id=true
+
+quarkus.hivemq.devservices.enabled=false
+%dev-service.quarkus.hivemq.devservices.enabled=true

--- a/integration-tests/hivemq-client-smallrye/src/test/java/io/quarkiverse/hivemqclient/test/smallrye/DevServiceNoAuthPriceTest.java
+++ b/integration-tests/hivemq-client-smallrye/src/test/java/io/quarkiverse/hivemqclient/test/smallrye/DevServiceNoAuthPriceTest.java
@@ -1,7 +1,10 @@
 package io.quarkiverse.hivemqclient.test.smallrye;
 
+import io.quarkiverse.hivemqclient.test.smallrye.resources.HivemqProfiles;
 import io.quarkus.test.junit.QuarkusTest;
+import io.quarkus.test.junit.TestProfile;
 
+@TestProfile(HivemqProfiles.DevServiceEnabled.class)
 @QuarkusTest
 public class DevServiceNoAuthPriceTest extends CommonScenarios {
 }

--- a/integration-tests/hivemq-client-smallrye/src/test/java/io/quarkiverse/hivemqclient/test/smallrye/resources/CommonResources.java
+++ b/integration-tests/hivemq-client-smallrye/src/test/java/io/quarkiverse/hivemqclient/test/smallrye/resources/CommonResources.java
@@ -19,6 +19,12 @@ public abstract class CommonResources implements QuarkusTestResourceLifecycleMan
         config.put("mp.messaging.outgoing.topic-price.port", "" + getHiveMQContainer().getMqttPort());
         config.put("mp.messaging.incoming.prices.host", getHiveMQContainer().getHost());
         config.put("mp.messaging.incoming.prices.port", "" + getHiveMQContainer().getMqttPort());
+
+        // custom topic scenario
+        config.put("mp.messaging.incoming.custom-topic-sink.host", getHiveMQContainer().getHost());
+        config.put("mp.messaging.incoming.custom-topic-sink.port", "" + getHiveMQContainer().getMqttPort());
+        config.put("mp.messaging.outgoing.custom-topic.host", getHiveMQContainer().getHost());
+        config.put("mp.messaging.outgoing.custom-topic.port", "" + getHiveMQContainer().getMqttPort());
         return config;
     }
 

--- a/integration-tests/hivemq-client-smallrye/src/test/java/io/quarkiverse/hivemqclient/test/smallrye/resources/HivemqProfiles.java
+++ b/integration-tests/hivemq-client-smallrye/src/test/java/io/quarkiverse/hivemqclient/test/smallrye/resources/HivemqProfiles.java
@@ -40,4 +40,11 @@ public class HivemqProfiles {
             return Collections.singletonList(new TestResourceEntry(EnterpriseJwtAuthResources.class));
         }
     }
+
+    public static class DevServiceEnabled implements QuarkusTestProfile {
+        @Override
+        public String getConfigProfile() {
+            return "dev-service";
+        }
+    }
 }

--- a/integration-tests/hivemq-client-smallrye/src/test/java/io/quarkiverse/hivemqclient/test/smallrye/resources/MtlsAuthResources.java
+++ b/integration-tests/hivemq-client-smallrye/src/test/java/io/quarkiverse/hivemqclient/test/smallrye/resources/MtlsAuthResources.java
@@ -48,6 +48,25 @@ public class MtlsAuthResources extends HiveMQCommunityEdition {
         config.put("mp.messaging.incoming.prices.ssl.truststore.password", "changeme");
         config.put("mp.messaging.incoming.prices.ssl.truststore.hostVerifier", "false");
 
+        config.put("mp.messaging.outgoing.custom-topic.ssl", "true");
+        config.put("mp.messaging.outgoing.custom-topic.ssl.keystore.type", "jks");
+        config.put("mp.messaging.outgoing.custom-topic.ssl.keystore.location", "src/test/resources/certs/client-keystore.jks");
+        config.put("mp.messaging.outgoing.custom-topic.ssl.keystore.password", "changeme");
+        config.put("mp.messaging.outgoing.custom-topic.ssl.truststore.type", "jks");
+        config.put("mp.messaging.outgoing.custom-topic.ssl.truststore.location", "src/test/resources/certs/server.jks");
+        config.put("mp.messaging.outgoing.custom-topic.ssl.truststore.password", "changeme");
+        config.put("mp.messaging.outgoing.custom-topic.ssl.truststore.hostVerifier", "false");
+
+        config.put("mp.messaging.incoming.custom-topic-sink.ssl", "true");
+        config.put("mp.messaging.incoming.custom-topic-sink.ssl.keystore.type", "jks");
+        config.put("mp.messaging.incoming.custom-topic-sink.ssl.keystore.location",
+                "src/test/resources/certs/client-keystore.jks");
+        config.put("mp.messaging.incoming.custom-topic-sink.ssl.keystore.password", "changeme");
+        config.put("mp.messaging.incoming.custom-topic-sink.ssl.truststore.type", "jks");
+        config.put("mp.messaging.incoming.custom-topic-sink.ssl.truststore.location", "src/test/resources/certs/server.jks");
+        config.put("mp.messaging.incoming.custom-topic-sink.ssl.truststore.password", "changeme");
+        config.put("mp.messaging.incoming.custom-topic-sink.ssl.truststore.hostVerifier", "false");
+
         return config;
     }
 }

--- a/integration-tests/hivemq-client-smallrye/src/test/java/io/quarkiverse/hivemqclient/test/smallrye/resources/RbacAuthResources.java
+++ b/integration-tests/hivemq-client-smallrye/src/test/java/io/quarkiverse/hivemqclient/test/smallrye/resources/RbacAuthResources.java
@@ -24,6 +24,10 @@ public class RbacAuthResources extends HiveMQCommunityEdition {
         config.put("mp.messaging.outgoing.topic-price.password", DEFAULT_PASSWORD);
         config.put("mp.messaging.incoming.prices.username", DEFAULT_USERNAME);
         config.put("mp.messaging.incoming.prices.password", DEFAULT_PASSWORD);
+        config.put("mp.messaging.incoming.custom-topic-sink.username", DEFAULT_USERNAME);
+        config.put("mp.messaging.incoming.custom-topic-sink.password", DEFAULT_PASSWORD);
+        config.put("mp.messaging.outgoing.custom-topic.username", DEFAULT_USERNAME);
+        config.put("mp.messaging.outgoing.custom-topic.password", DEFAULT_PASSWORD);
         return config;
     }
 }

--- a/integration-tests/hivemq-client-smallrye/src/test/java/io/quarkiverse/hivemqclient/test/smallrye/resources/TlsAuthResources.java
+++ b/integration-tests/hivemq-client-smallrye/src/test/java/io/quarkiverse/hivemqclient/test/smallrye/resources/TlsAuthResources.java
@@ -33,6 +33,17 @@ public class TlsAuthResources extends HiveMQCommunityEdition {
         config.put("mp.messaging.incoming.prices.ssl.truststore.password", "changeme");
         config.put("mp.messaging.incoming.prices.ssl.truststore.hostVerifier", "false");
 
+        config.put("mp.messaging.outgoing.custom-topic.ssl", "true");
+        config.put("mp.messaging.outgoing.custom-topic.ssl.truststore.type", "jks");
+        config.put("mp.messaging.outgoing.custom-topic.ssl.truststore.location", "src/test/resources/certs/server.jks");
+        config.put("mp.messaging.outgoing.custom-topic.ssl.truststore.password", "changeme");
+        config.put("mp.messaging.outgoing.custom-topic.ssl.truststore.hostVerifier", "false");
+
+        config.put("mp.messaging.incoming.custom-topic-sink.ssl", "true");
+        config.put("mp.messaging.incoming.custom-topic-sink.ssl.truststore.type", "jks");
+        config.put("mp.messaging.incoming.custom-topic-sink.ssl.truststore.location", "src/test/resources/certs/server.jks");
+        config.put("mp.messaging.incoming.custom-topic-sink.ssl.truststore.password", "changeme");
+        config.put("mp.messaging.incoming.custom-topic-sink.ssl.truststore.hostVerifier", "false");
         return config;
     }
 }

